clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -295,8 +299,11 @@ private void doSend(ClientRequest clientRequest, boolean isInternalRequest, long
                 if (discoverBrokerVersions && log.isTraceEnabled())
                     log.trace("No version information found when sending message of type {} to node {}. " +
                             "Assuming version {}.", clientRequest.apiKey(), nodeId, version);
-            } else {
+            } else if (builder.desiredVersion() == null) {
                 version = versionInfo.usableVersion(clientRequest.apiKey());
+            } else {
+                versionInfo.ensureUsable(clientRequest.apiKey(), builder.desiredVersion());
+                version = builder.desiredVersion();
```

Comment:
Hmm, would it be better to have a `NodeApiVersions.usableVersion` that takes a `desiredVersion` as well? We could the collapse these two branches and there would be no need to call `ensureUsable`.